lang/LangPrimSource/PyrUnixPrim.cpp

```diff
@@ -186,6 +185,79 @@ int prString_POpen(struct VMGlobals *g, int numArgsPushed)
 	return errNone;
 }
 
+int prArrayPOpen(struct VMGlobals *g, int numArgsPushed);
+int prArrayPOpen(struct VMGlobals *g, int numArgsPushed)
+{
+	PyrObject *obj;
+
+	PyrSlot *a = g->sp - 1;
+	PyrSlot *b = g->sp;
+	
+#ifdef SC_IPHONE
+	SetInt(a, 0);
+	return errNone;
+#endif
+	
+	if (NotObj(a)) return errWrongType;
+
+	obj = slotRawObject(a);
+	if (!(slotRawInt(&obj->classptr->classFlags) & classHasIndexableInstances))
+		return errNotAnIndexableObject;
+		
+	if( obj->size < 1)
+		return errFailed;
+		
+	PyrSlot filenameSlot;
+	getIndexedSlot(obj, &filenameSlot, 0);
+	if (!isKindOfSlot(&filenameSlot, class_string)) return errWrongType;
+	char filename[slotRawObject(&filenameSlot)->size + 1];
+	slotStrVal(&filenameSlot, filename, slotRawObject(&filenameSlot)->size + 1);
+	
+	std::vector<char *> argv (obj->size + 1);
+	
+	path p;
+	p /= filename;
+	std::string filenameOnly = p.filename().string();
+	std::vector<char> vfilenameOnly(filenameOnly.begin(), filenameOnly.end());
+	vfilenameOnly.push_back('\0');
+	
+	argv[0] = vfilenameOnly.data();
+	argv[obj->size] = NULL;
+		
+	if(obj->size > 1) {
+		for (int i=1; i<obj->size; ++i) {
+			PyrSlot argSlot;
+			getIndexedSlot(obj, &argSlot, i);
+			if (!isKindOfSlot(&argSlot, class_string)) return errWrongType;
+			char *arg = new char[slotRawObject(&argSlot)->size + 1];
+			slotStrVal(&argSlot, arg, slotRawObject(&argSlot)->size + 1);
+			argv[i] = arg;
+		}
+	}
+	
+	sc_process *process = new sc_process;
+	process->stream = sc_popen_argv(filename, argv.data(), &process->pid, "r");
+	setvbuf(process->stream, 0, _IONBF, 0);
+
+	process->postOutput = IsTrue(b);
+
+	if(process->stream == NULL) {
+		delete process;
+		return errFailed;
+	}
+
+	thread thread(std::bind(string_popen_thread_func, process));
+	thread.detach();
+
+	for (int i=1; i<obj->size; ++i) {
+		delete [] argv[i];
+	}
+
+	SetInt(a, process->pid);
```

Comment:
maybe a bit safer: don't dereference process after passing it to `string_popen_thread_func`: the thread might finish and deallocate the process struct before you read the pid